File: TwistedPair.java
/************************************************************
* 作者：伍耀晖              Author: Geek.Zhiyuan            *
* 开源日期：2022年6月7日    Open Source Date: 2022-6-7      *
* 国家城市：中国广州        City, Country: GuangZhou, China *
************************************************************/

import java.io.*;

public class TwistedPair
{
    private static void Usage()
    {
        System.out.println("Usage\n\tEncryption: java TwistedPair -e/-E Plaintext.file Ciphertext.file Password\n\tDecryption: java TwistedPair -d/-D Ciphertext.file Plaintext.file Password");
    }

    private static void Trinary(byte[] baTrinary,
                                int iNumeric)
    {
        baTrinary[0] = baTrinary[1] = baTrinary[2] = baTrinary[3] = baTrinary[4] = baTrinary[5] =  0;

        if (iNumeric != 0)
        {
            for (int i = 5; i > -1; --i)
            {
                baTrinary[i] = (byte)(iNumeric % 3 & 255);

                iNumeric /= 3;
            }
        }
    }

// 0 0 2
// 1 1 1
// 2 2 0

    private static void TrinaryXOr0(byte[] baPlaintextOrCiphertext,
                                    byte[] baPassword)
    {
        for (int i = 0; i < 6; ++i)
        {
            if (baPlaintextOrCiphertext[i] == 0 && baPassword[i] == 0)
            {
                baPlaintextOrCiphertext[i] = 0;
            }
            else if (baPlaintextOrCiphertext[i] == 0 && baPassword[i] == 1)
            {
                baPlaintextOrCiphertext[i] = 0;
            }
            else if (baPlaintextOrCiphertext[i] == 0 && baPassword[i] == 2)
            {
                baPlaintextOrCiphertext[i] = 2;
            }
            else if (baPlaintextOrCiphertext[i] == 1 && baPassword[i] == 0)
            {
                baPlaintextOrCiphertext[i] = 1;
            }
            else if (baPlaintextOrCiphertext[i] == 1 && baPassword[i] == 1)
            {
                baPlaintextOrCiphertext[i] = 1;
            }
            else if (baPlaintextOrCiphertext[i] == 1 && baPassword[i] == 2)
            {
                baPlaintextOrCiphertext[i] = 1;
            }
            else if (baPlaintextOrCiphertext[i] == 2 && baPassword[i] == 0)
            {
                baPlaintextOrCiphertext[i] = 2;
            }
            else if (baPlaintextOrCiphertext[i] == 2 && baPassword[i] == 1)
            {
                baPlaintextOrCiphertext[i] = 2;
            }
            else if (baPlaintextOrCiphertext[i] == 2 && baPassword[i] == 2)
            {
                baPlaintextOrCiphertext[i] = 0;
            }
        }
    }

// 0 2 2
// 1 1 1
// 2 0 0

    private static void TrinaryXOr2(byte[] baPlaintextOrCiphertext,
                                    byte[] baPassword)
    {
        for (int j = 0; j < 6; ++j)
        {
            if (baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 0)
            {
                baPlaintextOrCiphertext[j] = 0;
            }
            else if (baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 1)
            {
                baPlaintextOrCiphertext[j] = 2;
            }
            else if (baPlaintextOrCiphertext[j] == 0 && baPassword[j] == 2)
            {
                baPlaintextOrCiphertext[j] = 2;
            }
            else if (baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 0)
            {
                baPlaintextOrCiphertext[j] = 1;
            }
            else if (baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 1)
            {
                baPlaintextOrCiphertext[j] = 1;
            }
            else if (baPlaintextOrCiphertext[j] == 1 && baPassword[j] == 2)
            {
                baPlaintextOrCiphertext[j] = 1;
            }
            else if (baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 0)
            {
                baPlaintextOrCiphertext[j] = 2;
            }
            else if (baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 1)
            {
                baPlaintextOrCiphertext[j] = 0;
            }
            else if (baPlaintextOrCiphertext[j] == 2 && baPassword[j] == 2)
            {
                baPlaintextOrCiphertext[j] = 0;
            }
        }
    }

// 2 0 0
// 1 1 1
// 0 2 2

    private static void TrinaryXAnd0(byte[] baPlaintextOrCiphertext,
                                     byte[] baPassword)
    {
        for (int k = 0; k < 6; ++k)
        {
            if (baPlaintextOrCiphertext[k] == 0 && baPassword[k] == 0)
            {
                baPlaintextOrCiphertext[k] = 2;
            }
            else if (baPlaintextOrCiphertext[k] == 0 && baPassword[k] == 1)
            {
                baPlaintextOrCiphertext[k] = 0;
            }
            else if (baPlaintextOrCiphertext[k] == 0 && baPassword[k] == 2)
            {
                baPlaintextOrCiphertext[k] = 0;
            }
            else if (baPlaintextOrCiphertext[k] == 1 && baPassword[k] == 0)
            {
                baPlaintextOrCiphertext[k] = 1;
            }
            else if (baPlaintextOrCiphertext[k] == 1 && baPassword[k] == 1)
            {
                baPlaintextOrCiphertext[k] = 1;
            }
            else if (baPlaintextOrCiphertext[k] == 1 && baPassword[k] == 2)
            {
                baPlaintextOrCiphertext[k] = 1;
            }
            else if (baPlaintextOrCiphertext[k] == 2 && baPassword[k] == 0)
            {
                baPlaintextOrCiphertext[k] = 0;
            }
            else if (baPlaintextOrCiphertext[k] == 2 && baPassword[k] == 1)
            {
                baPlaintextOrCiphertext[k] = 2;
            }
            else if (baPlaintextOrCiphertext[k] == 2 && baPassword[k] == 2)
            {
                baPlaintextOrCiphertext[k] = 2;
            }
        }
    }

// 2 2 0
// 1 1 1
// 0 0 2

    private static void TrinaryXAnd2(byte[] baPlaintextOrCiphertext,
                                     byte[] baPassword)
    {
        for (int l = 0; l < 6; ++l)
        {
            if (baPlaintextOrCiphertext[l] == 0 && baPassword[l] == 0)
            {
                baPlaintextOrCiphertext[l] = 2;
            }
            else if (baPlaintextOrCiphertext[l] == 0 && baPassword[l] == 1)
            {
                baPlaintextOrCiphertext[l] = 2;
            }
            else if (baPlaintextOrCiphertext[l] == 0 && baPassword[l] == 2)
            {
                baPlaintextOrCiphertext[l] = 0;
            }
            else if (baPlaintextOrCiphertext[l] == 1 && baPassword[l] == 0)
            {
                baPlaintextOrCiphertext[l] = 1;
            }
            else if (baPlaintextOrCiphertext[l] == 1 && baPassword[l] == 1)
            {
                baPlaintextOrCiphertext[l] = 1;
            }
            else if (baPlaintextOrCiphertext[l] == 1 && baPassword[l] == 2)
            {
                baPlaintextOrCiphertext[l] = 1;
            }
            else if (baPlaintextOrCiphertext[l] == 2 && baPassword[l] == 0)
            {
                baPlaintextOrCiphertext[l] = 0;
            }
            else if (baPlaintextOrCiphertext[l] == 2 && baPassword[l] == 1)
            {
                baPlaintextOrCiphertext[l] = 0;
            }
            else if (baPlaintextOrCiphertext[l] == 2 && baPassword[l] == 2)
            {
                baPlaintextOrCiphertext[l] = 2;
            }
        }
    }

    public static void main(String[] args)
    {
        if (args.length != 4)
        {
            Usage();
        }
        else if (args[0].equals("-e") || args[0].equals("-E"))
        {
            RandomAccessFile rafPlaintext = null, rafCiphertext = null;

            try
            {
                rafPlaintext = new RandomAccessFile(args[1], "r");

                rafCiphertext = new RandomAccessFile(args[2], "rw");

                long lFileSize = rafPlaintext.length();

                if (lFileSize == 0)
                {
                    System.out.printf("There is no data in file [%s], 0 byte.\n", args[1]);

                    System.exit(-1);
                }

                rafCiphertext.setLength(2 * lFileSize);

                byte[] baPlaintextOrCiphertext = new byte[6];

                int iPasswordLength = args[3].length();

                byte[][] baPassword = new byte[iPasswordLength][6];

                for (int i = 0; i < iPasswordLength; ++i)
                {
                    Trinary(baPassword[i], args[3].getBytes()[i]);
                }

                int k = 0;

                for (long j = 0; j < lFileSize; ++j)
                {
                    Trinary(baPlaintextOrCiphertext, rafPlaintext.readUnsignedByte());

                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);


                    /*TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr0(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXOr2(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);


                    TrinaryXAnd2(baPlaintextOrCiphertext, baPassword[k]);

                    TrinaryXAnd0(baPlaintextOrCiphertext, baPassword[k]);*/

                    rafCiphertext.writeShort(243 * baPlaintextOrCiphertext[0] + 81 * baPlaintextOrCiphertext[1] + 27 * baPlaintextOrCiphertext[2] + 9 * baPlaintextOrCiphertext[3] + 3 * baPlaintextOrCiphertext[4] + baPlaintextOrCiphertext[5]);

                    k = ++k % iPasswordLength;
                }              
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
            finally
            {
                try
                {
                    if (rafCiphertext != null)
                    {
                        rafCiphertext.close();
                    }

                    if (rafPlaintext != null)
                    {
                        rafPlaintext.close();
                    }
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
        }
        else if (args[0].equals("-d") || args[0].equals("-D"))
        {
            RandomAccessFile rafCiphertext = null, rafPlaintext = null;

            try
            {
                rafCiphertext = new RandomAccessFile(args[1], "r");

                rafPlaintext = new RandomAccessFile(args[2], "rw");

                long lFileSize = rafCiphertext.length() / 2;

                if (lFileSize == 0)
                {
                    System.out.printf("There is no data in file [%s], 0 byte.\n", args[1]);

                    System.exit(-1);
                }

                rafPlaintext.setLength(lFileSize);

                byte[] baCiphertextOrPlaintext = new byte[6];

                int iPasswordLength = args[3].length();

                byte[][] baPassword = new byte[iPasswordLength][6];

                for (int i = 0; i < iPasswordLength; ++i)
                {
                    Trinary(baPassword[i], args[3].getBytes()[i]);
                }

                int k = 0;

                for (long j = 0; j < lFileSize; ++j)
                {
                    Trinary(baCiphertextOrPlaintext, rafCiphertext.readUnsignedShort());

                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);


                    /*TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXOr0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXOr2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);


                    TrinaryXAnd0(baCiphertextOrPlaintext, baPassword[k]);

                    TrinaryXAnd2(baCiphertextOrPlaintext, baPassword[k]);*/

                    rafPlaintext.writeByte(243 * baCiphertextOrPlaintext[0] + 81 * baCiphertextOrPlaintext[1] + 27 * baCiphertextOrPlaintext[2] + 9 * baCiphertextOrPlaintext[3] + 3 * baCiphertextOrPlaintext[4] + baCiphertextOrPlaintext[5]);

                    k = ++k % iPasswordLength;
                }
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
            finally
            {
                try
                {
                    if (rafPlaintext != null)
                    {
                        rafPlaintext.close();
                    }

                    if (rafCiphertext != null)
                    {
                        rafCiphertext.close();
                    }
                }
                catch (Exception e)
                {
                    e.printStackTrace();
                }
            }
        }
        else
        {
            Usage();
        }
    }
}

File: TwistedPair.c
/************************************************************
* 作者：伍耀晖              Author: Geek.Zhiyuan            *
* 开源日期：2022年6月7日    Open Source Date: 2022-6-7      *
* 国家城市：中国广州        City, Country: GuangZhou, China *
************************************************************/

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>

void Usage()
{
    printf("Usage\n\tEncryption: TwistedPair -e/-E Plaintext.file Ciphertext.file Password\n\tDecryption: TwistedPair -d/-D Ciphertext.file Plaintext.file Password\n");
}

void Trinary(unsigned char *ucpTrinary,
             long long lNumeric)
{
    ucpTrinary[0] = ucpTrinary[1] = ucpTrinary[2] = ucpTrinary[3] = ucpTrinary[4] = ucpTrinary[5] =  0;

    if (lNumeric != 0)
    {
        for (long long i = 5; i > -1; --i)
        {
            ucpTrinary[i] = lNumeric % 3;

            lNumeric /= 3;
        }
    }
}

// 0 0 2
// 1 1 1
// 2 2 0

void TrinaryXOr0(unsigned char *ucpPlaintextOrCiphertext,
                 unsigned char *ucpPassword)
{
    for (long long i = 0; i < 6; ++i)
    {
        if (ucpPlaintextOrCiphertext[i] == 0 && ucpPassword[i] == 0)
        {
            ucpPlaintextOrCiphertext[i] = 0;
        }
        else if (ucpPlaintextOrCiphertext[i] == 0 && ucpPassword[i] == 1)
        {
            ucpPlaintextOrCiphertext[i] = 0;
        }
        else if (ucpPlaintextOrCiphertext[i] == 0 && ucpPassword[i] == 2)
        {
            ucpPlaintextOrCiphertext[i] = 2;
        }
        else if (ucpPlaintextOrCiphertext[i] == 1 && ucpPassword[i] == 0)
        {
            ucpPlaintextOrCiphertext[i] = 1;
        }
        else if (ucpPlaintextOrCiphertext[i] == 1 && ucpPassword[i] == 1)
        {
            ucpPlaintextOrCiphertext[i] = 1;
        }
        else if (ucpPlaintextOrCiphertext[i] == 1 && ucpPassword[i] == 2)
        {
            ucpPlaintextOrCiphertext[i] = 1;
        }
        else if (ucpPlaintextOrCiphertext[i] == 2 && ucpPassword[i] == 0)
        {
            ucpPlaintextOrCiphertext[i] = 2;
        }
        else if (ucpPlaintextOrCiphertext[i] == 2 && ucpPassword[i] == 1)
        {
            ucpPlaintextOrCiphertext[i] = 2;
        }
        else if (ucpPlaintextOrCiphertext[i] == 2 && ucpPassword[i] == 2)
        {
            ucpPlaintextOrCiphertext[i] = 0;
        }
    }
}

// 0 2 2
// 1 1 1
// 2 0 0

void TrinaryXOr2(unsigned char *ucpPlaintextOrCiphertext,
                 unsigned char *ucpPassword)
{
    for (long long j = 0; j < 6; ++j)
    {
        if (ucpPlaintextOrCiphertext[j] == 0 && ucpPassword[j] == 0)
        {
            ucpPlaintextOrCiphertext[j] = 0;
        }
        else if (ucpPlaintextOrCiphertext[j] == 0 && ucpPassword[j] == 1)
        {
            ucpPlaintextOrCiphertext[j] = 2;
        }
        else if (ucpPlaintextOrCiphertext[j] == 0 && ucpPassword[j] == 2)
        {
            ucpPlaintextOrCiphertext[j] = 2;
        }
        else if (ucpPlaintextOrCiphertext[j] == 1 && ucpPassword[j] == 0)
        {
            ucpPlaintextOrCiphertext[j] = 1;
        }
        else if (ucpPlaintextOrCiphertext[j] == 1 && ucpPassword[j] == 1)
        {
            ucpPlaintextOrCiphertext[j] = 1;
        }
        else if (ucpPlaintextOrCiphertext[j] == 1 && ucpPassword[j] == 2)
        {
            ucpPlaintextOrCiphertext[j] = 1;
        }
        else if (ucpPlaintextOrCiphertext[j] == 2 && ucpPassword[j] == 0)
        {
            ucpPlaintextOrCiphertext[j] = 2;
        }
        else if (ucpPlaintextOrCiphertext[j] == 2 && ucpPassword[j] == 1)
        {
            ucpPlaintextOrCiphertext[j] = 0;
        }
        else if (ucpPlaintextOrCiphertext[j] == 2 && ucpPassword[j] == 2)
        {
            ucpPlaintextOrCiphertext[j] = 0;
        }
    }
}

// 2 0 0
// 1 1 1
// 0 2 2

void TrinaryXAnd0(unsigned char *ucpPlaintextOrCiphertext,
                  unsigned char *ucpPassword)
{
    for (long long k = 0; k < 6; ++k)
    {
        if (ucpPlaintextOrCiphertext[k] == 0 && ucpPassword[k] == 0)
        {
            ucpPlaintextOrCiphertext[k] = 2;
        }
        else if (ucpPlaintextOrCiphertext[k] == 0 && ucpPassword[k] == 1)
        {
            ucpPlaintextOrCiphertext[k] = 0;
        }
        else if (ucpPlaintextOrCiphertext[k] == 0 && ucpPassword[k] == 2)
        {
            ucpPlaintextOrCiphertext[k] = 0;
        }
        else if (ucpPlaintextOrCiphertext[k] == 1 && ucpPassword[k] == 0)
        {
            ucpPlaintextOrCiphertext[k] = 1;
        }
        else if (ucpPlaintextOrCiphertext[k] == 1 && ucpPassword[k] == 1)
        {
            ucpPlaintextOrCiphertext[k] = 1;
        }
        else if (ucpPlaintextOrCiphertext[k] == 1 && ucpPassword[k] == 2)
        {
            ucpPlaintextOrCiphertext[k] = 1;
        }
        else if (ucpPlaintextOrCiphertext[k] == 2 && ucpPassword[k] == 0)
        {
            ucpPlaintextOrCiphertext[k] = 0;
        }
        else if (ucpPlaintextOrCiphertext[k] == 2 && ucpPassword[k] == 1)
        {
            ucpPlaintextOrCiphertext[k] = 2;
        }
        else if (ucpPlaintextOrCiphertext[k] == 2 && ucpPassword[k] == 2)
        {
            ucpPlaintextOrCiphertext[k] = 2;
        }
    }
}

// 2 2 0
// 1 1 1
// 0 0 2

void TrinaryXAnd2(unsigned char *ucpPlaintextOrCiphertext,
                  unsigned char *ucpPassword)
{
    for (long long l = 0; l < 6; ++l)
    {
        if (ucpPlaintextOrCiphertext[l] == 0 && ucpPassword[l] == 0)
        {
            ucpPlaintextOrCiphertext[l] = 2;
        }
        else if (ucpPlaintextOrCiphertext[l] == 0 && ucpPassword[l] == 1)
        {
            ucpPlaintextOrCiphertext[l] = 2;
        }
        else if (ucpPlaintextOrCiphertext[l] == 0 && ucpPassword[l] == 2)
        {
            ucpPlaintextOrCiphertext[l] = 0;
        }
        else if (ucpPlaintextOrCiphertext[l] == 1 && ucpPassword[l] == 0)
        {
            ucpPlaintextOrCiphertext[l] = 1;
        }
        else if (ucpPlaintextOrCiphertext[l] == 1 && ucpPassword[l] == 1)
        {
            ucpPlaintextOrCiphertext[l] = 1;
        }
        else if (ucpPlaintextOrCiphertext[l] == 1 && ucpPassword[l] == 2)
        {
            ucpPlaintextOrCiphertext[l] = 1;
        }
        else if (ucpPlaintextOrCiphertext[l] == 2 && ucpPassword[l] == 0)
        {
            ucpPlaintextOrCiphertext[l] = 0;
        }
        else if (ucpPlaintextOrCiphertext[l] == 2 && ucpPassword[l] == 1)
        {
            ucpPlaintextOrCiphertext[l] = 0;
        }
        else if (ucpPlaintextOrCiphertext[l] == 2 && ucpPassword[l] == 2)
        {
            ucpPlaintextOrCiphertext[l] = 2;
        }
    }
}

long long main(long long argc,
               char *argv[])
{
    if (argc != 5)
    {
        Usage();
    }
    else if (*(short*)argv[1] == 0x452D || *(short*)argv[1] == 0x652D)
    {
        struct stat tStatFileSize;

        stat(argv[2], &tStatFileSize);

        long long lFileSize = tStatFileSize.st_size;

        if (lFileSize == 0)
        {
            printf("There is no data in file [%s], 0 byte.\n", argv[2]);

            return -1;
        }

        int fdPlaintextOrCiphertext = open(argv[2], O_BINARY | O_RDONLY, S_IREAD | S_IWRITE);

        unsigned char *ucpPlaintext = (unsigned char*)malloc(lFileSize), ucaPlaintextOrCiphertext[6];

        read(fdPlaintextOrCiphertext, ucpPlaintext, lFileSize);

        close(fdPlaintextOrCiphertext);

        unsigned short *uspCiphertext = (unsigned short*)malloc(2 * lFileSize);

        unsigned char ucPasswordLength = -1;

        while(argv[4][++ucPasswordLength]);

        unsigned char *ucpPassword = (unsigned char*)malloc(6 * ucPasswordLength);

        for (unsigned char i = 0; i < ucPasswordLength; ++i)
        {
            Trinary(ucpPassword + 6 * i, argv[4][i]);
        }

        for (unsigned long long j = 0, k = 0; j < lFileSize; ++j)
        {
            Trinary(ucaPlaintextOrCiphertext, ucpPlaintext[j]);

            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            /*TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaPlaintextOrCiphertext, ucpPassword + 6 * k);*/

            uspCiphertext[j] = 243 * ucaPlaintextOrCiphertext[0] + 81 * ucaPlaintextOrCiphertext[1] + 27 * ucaPlaintextOrCiphertext[2] + 9 * ucaPlaintextOrCiphertext[3] + 3 * ucaPlaintextOrCiphertext[4] + ucaPlaintextOrCiphertext[5];

            k = ++k % ucPasswordLength;
        }

        fdPlaintextOrCiphertext = open(argv[3], O_BINARY | O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);

        write(fdPlaintextOrCiphertext, uspCiphertext, 2 * lFileSize);

        close(fdPlaintextOrCiphertext);

        free(uspCiphertext);

        free(ucpPlaintext);
    }
    else if (*(short*)argv[1] == 0x442D || *(short*)argv[1] == 0x642D)
    {
        struct stat tStatFileSize;

        stat(argv[2], &tStatFileSize);

        long long lFileSize = tStatFileSize.st_size;

        if (lFileSize == 0)
        {
            printf("There is no data in file [%s], 0 byte.\n", argv[2]);

            return -1;
        }

        int fdCiphertextOrPlaintext = open(argv[2], O_BINARY | O_RDONLY, S_IREAD | S_IWRITE);

        unsigned short *uspCiphertext = (unsigned short*)malloc(lFileSize);

        read(fdCiphertextOrPlaintext, uspCiphertext, lFileSize);

        close(fdCiphertextOrPlaintext);

        lFileSize /= 2;

        unsigned char *ucpPlaintext = (unsigned char*)malloc(lFileSize), ucaCiphertextOrPlaintext[6];

        unsigned char ucPasswordLength = -1;

        while(argv[4][++ucPasswordLength]);

        unsigned char *ucpPassword = (unsigned char*)malloc(6 * ucPasswordLength);

        for (unsigned char i = 0; i < ucPasswordLength; ++i)
        {
            Trinary(ucpPassword + 6 * i, argv[4][i]);
        }

        for (unsigned long long j = 0, k = 0; j < lFileSize; ++j)
        {
            Trinary(ucaCiphertextOrPlaintext, uspCiphertext[j]);

            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            /*TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXOr0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXOr2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);


            TrinaryXAnd0(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);

            TrinaryXAnd2(ucaCiphertextOrPlaintext, ucpPassword + 6 * k);*/

            ucpPlaintext[j] = 243 * ucaCiphertextOrPlaintext[0] + 81 * ucaCiphertextOrPlaintext[1] + 27 * ucaCiphertextOrPlaintext[2] + 9 * ucaCiphertextOrPlaintext[3] + 3 * ucaCiphertextOrPlaintext[4] + ucaCiphertextOrPlaintext[5];

            k = ++k % ucPasswordLength;
        }

        fdCiphertextOrPlaintext = open(argv[3],  O_BINARY | O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);

        write(fdCiphertextOrPlaintext, ucpPlaintext, lFileSize);

        close(fdCiphertextOrPlaintext);

        free(ucpPlaintext);

        free(uspCiphertext);
    }
    else
    {
        Usage();
    }

    return 0;
}